### ISSUE TITLE
CAL & VAAL sampling not better than random

### ISSUE THREAD

**Issue by AronDiSc** [Sebi1128/DeepLearningProject#7]
Possible origins of error

**Embedding**:
1. Maybe encoder / bottleneck / decoder is wrongly parametrized (learning rate / loss weighting)
2. Maybe optimizer does not update parameters correctly

- [ ] Verify again if `vaal_with_latent` as the classifier really doesn't yield any improvements in terms of classifier accuracy
- [ ]  Verify if embedding works (e.g. train classifier only on latent variable `z` or try to visualize latent space with [`tsne`](https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html) e.g. in `test_epoch` in [training.py](https://github.com/Sebi1128/DeepLearningProject/blob/main/src/training.py))
- [ ] Check if better results when using [pca](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html) compared to vae

**Classifier**:
1. Maybe it is because the classifier overfits
- [ ] Change classifier optimizer setting (use smaller learning rate) and check validation loss -> when it doesn't overfit we check again the different sampling procedures

**Sampler**:
1. Maybe sampling process doesn't work
2. Maybe discriminator is trained incorrectly
3. Maybe cal sampling does not work at all?

- [ ] Check what happens when `sampler.sample` is called in [main.py](https://github.com/Sebi1128/DeepLearningProject/blob/main/main.py) when sampler is set to vaal or cal, does the sampling really work there?
- [ ] Check cal sampling by using it on a low dimensional representation (maybe possible by setting `z_dim: 2`) and visualize sampled results
- [ ] Check vaal sampling learning rate and optimizer. Is discriminator and generator validation loss decreasing?
- [ ] Check and adapt vaal sampling procedure in [training.py](https://github.com/Sebi1128/DeepLearningProject/blob/main/src/training.py). At the moment the standard approach for fan training is followed and not how it is done in the vaal paper. Maybe we can change the procedure s.t. it exactly matches how it is done in the vaal paper. Then we should see an improvement!
- [ ] Check also vaal sampling on low dimensional latent space. Visualize which points (e.g. in 2d) are sampled

**Comment by AronDiSc** [comment]
- [x] Maybe `x_save = x` on Line 40 in [model.py](https://github.com/Sebi1128/DeepLearningProject/blob/main/src/model.py) should be done with a copy?

**Comment by AronDiSc** [comment]
- [x] Add Kaiming init as in [vaal](https://github.com/sinhasam/vaal/blob/master/model.py#L111) line
- [ ] (Make VAE -> AE by setting KLD weight (beta?) = 0 and [`z=mu`](https://github.com/Sebi1128/DeepLearningProject/blob/main/src/base_models/bottlenecks.py#L68) instead of reparametrize in bottleneck)
- [x] [`eps = torch.rand_like(std)`](https://github.com/Sebi1128/DeepLearningProject/blob/main/src/base_models/bottlenecks.py#L59) should actually be `torch.randn(...)` because rand_like gives a uniform distribution but we want a normal distribution ([torch.rand_like](https://pytorch.org/docs/stable/generated/torch.rand_like.html))
- [x] [`c = self.classifier(x_save, z, x)`](https://github.com/Sebi1128/DeepLearningProject/blob/main/src/model.py#L52) to `c = self.classifier(x_save, latent[1], x)` because we want mu and not z (we want the max. likelihood est.)
- [ ] Big problem in sampling:

**Comment by AronDiSc** [comment]
> * [x]  Add Kaiming init as in [vaal](https://github.com/sinhasam/vaal/blob/master/model.py#L111) line
> * [ ]  (Make VAE -> AE by setting KLD weight (beta?) = 0 and [`z=mu`](https://github.com/Sebi1128/DeepLearningProject/blob/main/src/base_models/bottlenecks.py#L68) instead of reparametrize in bottleneck)
> * [x]  [`eps = torch.rand_like(std)`](https://github.com/Sebi1128/DeepLearningProject/blob/main/src/base_models/bottlenecks.py#L59) should actually be `torch.randn(...)` because rand_like gives a uniform distribution but we want a normal distribution ([torch.rand_like](https://pytorch.org/docs/stable/generated/torch.rand_like.html))
> * [x]  [`c = self.classifier(x_save, z, x)`](https://github.com/Sebi1128/DeepLearningProject/blob/main/src/model.py#L52) to `c = self.classifier(x_save, latent[1], x)` because we want mu and not z (we want the max. likelihood est.)
> * [x]  Big problem in sampling:
